clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -298,63 +357,43 @@ protected void onJoinComplete(int generation,
-                assignAndRevoke(listener, assignedPartitions, ownedPartitions);
+                Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
+                revokedPartitions.removeAll(assignedPartitions);
 
-                break;
-        }
+                log.info("Updating with newly assigned partitions: {}, compare with already owned partitions: {}, " +
+                        "newly added partitions: {}, revoking partitions: {}",
+                    Utils.join(assignedPartitions, ", "),
+                    Utils.join(ownedPartitions, ", "),
+                    Utils.join(addedPartitions, ", "),
+                    Utils.join(revokedPartitions, ", "));
 
-    }
+                // add partitions that were not previously owned but are now assigned
+                firstException.compareAndSet(null, maybeInvokePartitionsAssigned(addedPartitions));
 
-    private void assignAndRevoke(final ConsumerRebalanceListener listener,
-                                 final Set<TopicPartition> assignedPartitions,
-                                 final Set<TopicPartition> ownedPartitions) {
-        Set<TopicPartition> addedPartitions = new HashSet<>(assignedPartitions);
-        Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
-        addedPartitions.removeAll(ownedPartitions);
-        revokedPartitions.removeAll(assignedPartitions);
+                // revoked partitions that was previously owned but no longer assigned
+                firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
 
-        log.info("Updating with newly assigned partitions: {}, compare with already owned partitions: {}, " +
-                "newly added partitions: {}, revoking partitions: {}",
-            Utils.join(assignedPartitions, ", "),
-            Utils.join(ownedPartitions, ", "),
-            Utils.join(addedPartitions, ", "),
-            Utils.join(revokedPartitions, ", "));
+                // if revoked any partitions, need to re-join the group afterwards
+                if (!revokedPartitions.isEmpty()) {
+                    requestRejoin();
+                }
 
-        try {
-            listener.onPartitionsAssigned(addedPartitions);
-        } catch (WakeupException | InterruptException e) {
-            throw e;
-        } catch (Exception e) {
-            log.error("User provided listener {} failed on partition assignment", listener.getClass().getName(), e);
+                break;
         }
 
-        try {
-            listener.onPartitionsRevoked(revokedPartitions);
-        } catch (WakeupException | InterruptException e) {
-            throw e;
-        } catch (Exception e) {
-            log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
-        }
+        if (firstException.get() != null)
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
```

Comment:
nit: "User rebalance callback **threw** an error"

Reply:
Also, might it be good to clarify that this was the first exception but another method may also have thrown? Or even save the message from any/all exceptions thrown from callbacks?